experimental/dds/tree2/src/util/memoizedIdRangeAllocator.ts

```diff
@@ -0,0 +1,92 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { RangeMap, getFirstFromRangeMap, setInRangeMap } from "./rangeMap";
+import { Mutable } from "./utils";
+
+/**
+ * A value used to differentiate ranges of IDs that belong to different spaces.
+ * @alpha
+ */
+type SpaceKey = string | number | undefined;
+
+/**
+ * An unique ID allocator that returns the output ID for the same input ID.
+ * "The same" here includes cases where a prior call allocated a range of IDs that partially or fully overlap with the
+ * current call.
+ * @alpha
+ */
+export type MemoizedIdRangeAllocator = (
+	key: SpaceKey,
+	startId: number,
+	count?: number,
+) => IdRange[];
+
+/**
+ * @alpha
+ */
+export interface IdRange {
+	readonly first: number;
+	readonly count: number;
+}
+
+export const MemoizedIdRangeAllocator = {
+	fromNextId(nextId: number = 0): MemoizedIdRangeAllocator {
+		let _nextId = nextId;
+		const rangeMap: Map<SpaceKey, RangeMap<number>> = new Map();
+		return (key: string | number | undefined, startId: number, length?: number): IdRange[] => {
+			let count = length ?? 1;
+			const out: IdRange[] = [];
+			let ranges = rangeMap.get(key);
+			if (ranges === undefined) {
+				ranges = [];
+				rangeMap.set(key, ranges);
+			}
+			let currId = startId;
+			while (count > 0) {
+				const firstRange = getFirstFromRangeMap(ranges, currId, count);
```

Comment:
Not a big deal, but it might be nice to add RangeMap utilities to make this easier.

Reply:
What do you have in mind? A `getFirstOrAddFromRangeMap` would be awkward to use here because we need to know whether the _nextId needs incrementing.
Same for the parts where the range is extended.
I guess we could make it so that you don't get a bigger range than what you asked for. I'll try that.

Reply:
Actually, I don't like that, because in one case the range in the array is returned, and in the other (the trim case) a (modifed) copy of the range is returned. That's really confusing if you're updating the returned value.